File: Projetos_C/Terceiro Semestre/Estrutura de Dados/Codificação de Huffman/Huffman.c
#include <stdio.h>
#include <stdlib.h>
#define TAMANHO 27

struct No {
    unsigned char carac;
    int freq;
    struct No *esq, *dir, *pai;
};
typedef struct No *p_no;
typedef struct No no;

struct Pilha {
    int topo;
    p_no vetor_no[TAMANHO];
};
typedef struct Pilha *pilha_no;
typedef struct Pilha aux_no;

p_no criar_no(int freq, char carac, p_no esq, p_no dir) {
    p_no r = (p_no)malloc(sizeof(no));
    r->freq = freq;
    r->carac = carac;
    r->esq = esq;
    r->dir = dir;
    r->pai = NULL;

    if (r->esq != NULL)
        r->esq->pai = r;
    if (r->dir != NULL)
        r->dir->pai = r;
    return r;
}

void inicializa_vetor(int *vetor_freq) {
    for (int i = 0; i < TAMANHO; i++) {
        vetor_freq[i] = 0;
    }
}

int informa_indice(char caracter) {
    if (caracter == ' ') {
        return 0;
    } else {
        int valor = (caracter - 'a') + 1;
        return valor;
    }
}

void informa_freq(unsigned char *texto, int *vetor_freq) {
    int indice = 0;
    while (texto[indice] != '\0') {
        unsigned char caracter = texto[indice];
        int posicao = informa_indice(caracter);
        vetor_freq[posicao] += 1;
        indice++;
    }
}

void imprime_freq(int *vetor_freq) {
    for (int i = 0; i < TAMANHO; i++) {
        if (i == 0) {
            printf("  : %d\n", vetor_freq[i]);
        } else {
            printf("%c : %d\n", 'a' + (i - 1), vetor_freq[i]);
        }
    }
}

void push(pilha_no pilha, p_no novo_no) {
    if (pilha->topo == TAMANHO - 1)
        return;
    else {
        pilha->topo++;
        pilha->vetor_no[pilha->topo] = novo_no;
    }
}

void cria_pilha(pilha_no pilha, int *vetor_freq) {
    for (int i = 0; i < TAMANHO; i++) {
        if (vetor_freq[i] != 0) {
            unsigned char aux_caracter = 'a' + (i - 1);
            if (i == 0) {
                aux_caracter = ' ';
                p_no novo_no = criar_no(vetor_freq[i], aux_caracter, NULL, NULL);
                push(pilha, novo_no);
            } else {
                p_no novo_no = criar_no(vetor_freq[i], aux_caracter, NULL, NULL);
                push(pilha, novo_no);
            }
        }
    }
}

void Pop(pilha_no pilha) {
    if (pilha->topo == -1)
        return;
    else {
        pilha->topo--;
    }
}

void ordena_vetor(pilha_no pilha) {
    if (pilha->topo <= 0)
        return;

    for (int i = 0; i <= pilha->topo; i++) {
        for (int j = 0; j <= pilha->topo - 1 - i; j++) {
            if (pilha->vetor_no[j]->freq == pilha->vetor_no[j + 1]->freq) {
                if (pilha->vetor_no[j]->carac > pilha->vetor_no[j + 1]->carac) {
                    p_no temp = pilha->vetor_no[j];
                    pilha->vetor_no[j] = pilha->vetor_no[j + 1];
                    pilha->vetor_no[j + 1] = temp;
                }
            } else if (pilha->vetor_no[j]->freq < pilha->vetor_no[j + 1]->freq) {
                p_no temp = pilha->vetor_no[j];
                pilha->vetor_no[j] = pilha->vetor_no[j + 1];
                pilha->vetor_no[j + 1] = temp;
            }
        }
    }
}

p_no cria_arvore(pilha_no pilha) {
    while (pilha->topo > 0) {
        p_no aux1_no = pilha->vetor_no[pilha->topo];
        Pop(pilha);

        p_no aux2_no = pilha->vetor_no[pilha->topo];
        Pop(pilha);

        int soma_freq = (aux1_no->freq + aux2_no->freq);
        p_no novo_no = criar_no(soma_freq, '+', aux1_no, aux2_no);

        push(pilha, novo_no);
        //ordena_vetor(pilha);
    }
    return pilha->vetor_no[pilha->topo];
}

void imprimeArvore(p_no raiz) {
    if (raiz == NULL) {
        printf("Arvore vazia\n");
        return;
    }
    if (raiz->esq != NULL) {
        imprimeArvore(raiz->esq);
    }
    printf("%c ", raiz->carac);
    if (raiz->dir != NULL) {
        imprimeArvore(raiz->dir);
    }
}

int main(void) {
    pilha_no pilha = (pilha_no)malloc(sizeof(aux_no));
    pilha->topo = -1;
    unsigned char palavra[100];
    int vetor_freq[TAMANHO];
    inicializa_vetor(vetor_freq);
    informa_freq("", vetor_freq);
    imprime_freq(vetor_freq);

    cria_pilha(pilha, vetor_freq);
    ordena_vetor(pilha);

    for (int i = 0; i <= pilha->topo; i++) {
        unsigned char letra = pilha->vetor_no[i]->carac;
        printf("\n%c", letra);
    }

    p_no raiz = cria_arvore(pilha);
    printf("\nIn-order traversal of Huffman Tree:\n");
    imprimeArvore(raiz);
    printf("\n\n%c", raiz->esq->carac);
    printf("\n\n%c", raiz->dir->carac);
    printf("\n\n%c", raiz->esq->esq->carac);
    printf("\n\n%c", raiz->esq->dir->carac);
    free(pilha);
    return 0;
}
    

File: Projetos_C/Terceiro Semestre/Estrutura de Dados/Codificação de Huffman/Huffman.py
from Impressao import Desenho
TAMANHO = 27

class No:
    def __init__(self, freq, carac, esquerda=None, direita=None):
        self.freq = freq
        self.carac = carac
        self.esquerda = esquerda
        self.direita = direita
        self.pai = None
        
        # Atualiza ponteiro de pai se os filhos não forem nulos
        if self.esquerda:
            self.esquerda.pai = self
        if self.direita:
            self.direita.pai = self

class Frequencia:
    def __init__(self):
        self.vetor_freq = [0] * TAMANHO

    @staticmethod
    def informa_indice(caracter):
        if caracter == ' ':
            return 0
        else:
            valor = ord(caracter) - ord('a') + 1
            return valor
        
    def informa_freq(self, texto):
        for caracter in texto:
            posicao = Frequencia.informa_indice(caracter)
            self.vetor_freq[posicao] += 1

    def imprime_freq(self): 
        for i in range(TAMANHO): 
            if i == 0: 
                print("  : ", self.vetor_freq[i])
            else:
                print(f"{chr(ord('a') + i - 1)} : {self.vetor_freq[i]}")

class Arvore:
    def __init__(self):
        self.raiz = None
        self.lista = []
    
    def criar_lista(self, vetor_freq):
        for i in range(TAMANHO):
            if vetor_freq[i] != 0:
                carac = ' ' if i == 0 else chr(ord('a') + i - 1)
                novo_no = No(vetor_freq[i], carac)
                self.lista.append(novo_no)
        self.ordenar_lista()

    def ordenar_lista(self):
        self.lista.sort(key=lambda no: (no.freq, no.carac))

    def cria_arvore(self):
        while len(self.lista) > 1:
            # Remove os dois nós com menor frequência
            aux1_no = self.lista.pop(0)
            aux2_no = self.lista.pop(0)
            # Cria um novo nó combinando os dois removidos
            soma_freq = aux1_no.freq + aux2_no.freq
            novo_no = None
            novo_no = No(soma_freq, '|', aux1_no, aux2_no)
            #novo_no = No(soma_freq, '|', aux2_no, aux1_no)
            """
            if aux1_no.freq<=aux2_no.freq:
                novo_no = No(soma_freq, '|', aux1_no, aux2_no)
            else:
                novo_no = No(soma_freq, '|', aux2_no, aux1_no)
            """
            # Adiciona o novo nó de volta na lista
            self.lista.append(novo_no)
            self.ordenar_lista()
        
        # Define o nó raiz da árvore
        self.raiz = self.lista.pop()

    def imprimir_arvore(self, no, prefixo=''):
        if no is not None:
            print(f"{prefixo}Caractere: {no.carac}, Frequência: {no.freq}")
            self.imprimir_arvore(no.esquerda, prefixo + 'L-')
            self.imprimir_arvore(no.direita, prefixo + 'R-')
    
    def procurar_no(self, raiz, x):
        if raiz is None:
            return None

        if raiz.carac == x:
            return raiz

        left = self.procurar_no(raiz.esquerda, x)
        if left is not None:
            return left

        right = self.procurar_no(raiz.direita, x)
        if right is not None:
            return right

        return None

    def codifica_caractere(self, c):
        aux=self.procurar_no(self.raiz, c)
        cod=[]

        while aux.pai!=None:
            if aux.pai.esquerda == aux:
                cod.insert(0,'0')
            else:
                cod.insert(0,'1')
            aux=aux.pai
        return cod
    
    def decodifica(self, codificacao):
        decodificacao=[]
        if self.raiz != None:
            aux=self.raiz
            tam=len(codificacao)
            i=0
            while i<tam:
                    if aux.esquerda == None and aux.direita == None:
                        decodificacao.append(aux.carac)
                        aux=self.raiz
                    else:
                        if codificacao[i] == '0':
                            aux=aux.esquerda
                            i+=1
                        else:
                            aux=aux.direita
                            i+=1
            decodificacao.append(aux.carac)
        return decodificacao
            
    def codifica(self, texto):
        codificacao=[]
        for i in texto:
           codificacao.extend(self.codifica_caractere(i))
        return codificacao             
    
# Exemplo de uso
texto = "upa pepipa popipa papou pa papa e pipe ap"
freq = Frequencia()
freq.informa_freq(texto)
arvore = Arvore()
arvore.criar_lista(freq.vetor_freq)
arvore.cria_arvore()
codificacao=arvore.codifica(texto)
print('etapa 4 (extra):')
print()

print('original: "' + texto + '"')
print('\ncodificado:\n')

for i in codificacao:
    print(i, end='')

print('\n\n')

print('decodificado: "', end='')
decodificacao=arvore.decodifica(codificacao)
for i in decodificacao:
    print(i, end='')
print('"')

print('\n')

Desenho.desenhar(arvore.raiz)